[Feature:TAGrading]Peer_feedback_backend_implementation

jchulton opened the pull request:
### Please check if the PR fulfills these requirements:

* [ ] The PR title and message follows our [guidelines](https://submitty.org/developer/how_to_contribute)
* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
There is a peer feedback interface, but it doesn't actually save user input to the backend: it is just a UI
### What is the new behavior?
A student can give feedback via the UI, and a peer grader can view that feedback via the TA_Grading page.

Steps to use:
1. View grdaes of a student who has been peer graded
2. Use the Switch to Feedback Display button to switch to the feedback UI
3. Select a peer tab to give feedback to
4. Select the feedback to give in the peer feedback bar
5. Switch to the peer who have just given feedback to
6. Go to the TA_Grading page in which you graded that student who gave feedback. You should see the feedback at the bottom of the grading rubric

![image](https://user-images.githubusercontent.com/11095297/84407890-6e991300-abd9-11ea-979d-1fca9f4add97.png)


### Other information?
<!-- Is this a breaking change? -->
<!-- How did you test -->

jchulton commented:
Added the missing migration file

MasterOdin commented on site/app/controllers/grading/ElectronicGraderController.php:
```diff
@@ -2373,4 +2373,23 @@ private function getStats(Gradeable $gradeable, User $grader, bool $full_stats,
     }
+    
+    /**
+     * @Route("/{_semester}/{_course}/gradeable/{gradeable_id}/feedback/set", methods={"POST"})
```
```suggestion
     * @Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/feedback/set", methods={"POST"})
```

jchulton replied:
Currently, this change breaks the branch: buildCourseUrl in server.js does not include the /courses/.  After testing, before merge, I will make this change.

emaicus commented on migration/migrator/migrations/course/20200608034200_peer_feedback.py:
```diff
@@ -0,0 +1,105 @@
+    database.execute(
+        """
+       CREATE TABLE IF NOT EXISTS peer_feedback (
+            uid integer NOT NULL,
```
What is `uid` here?

emaicus replied:
I think we tend to name unique id's based on the table. For example, this would be `pf_id` for "peer feedback id"

jchulton replied:
uid has been changed to pf_id

emaicus commented on migration/migrator/migrations/course/20200608034200_peer_feedback.py:
```diff
@@ -0,0 +1,105 @@
+            team_id character varying(255),
+            g_id character varying(255) NOT NULL,
+            feedback_full character varying(255),
+            feedback_id character varying(255),
```
What is `feedback_id`?

emaicus commented on migration/migrator/migrations/course/20200608034200_peer_feedback.py:
```diff
@@ -0,0 +1,105 @@
+            user_id character varying(255),
+            team_id character varying(255),
+            g_id character varying(255) NOT NULL,
+            feedback_full character varying(255),
```
Why is this `feedback_full`? Is there a `feedback_short`?

jchulton replied:
feedback_id is what the sytem uses internally to id all the different feedbacks. feedback_full is what users see as feedback. This allows for ids to maintain standards regardless of the actual feedback

emaicus commented on site/app/controllers/grading/ElectronicGraderController.php:
```diff
@@ -2373,4 +2373,23 @@ private function getStats(Gradeable $gradeable, User $grader, bool $full_stats,
+    /**
+     * @Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/feedback/set", methods={"POST"})
+     */
+    public function setPeerFeedback($gradeable_id) {
```
If this is an ajax endpoint, rename this function `ajaxSerPeerFeedback` and add some access control.

jchulton replied:
Should there be access control? Everone should be able to give peer feedback, or am I misunderstanding access control?

jchulton replied:
I have changed to ajaxSetPeerFeedback

emaicus commented on site/app/controllers/pdf/PDFController.php:
```diff
@@ -20,7 +20,7 @@ public function __construct(Core $core) {
      * @Route("/{_semester}/{_course}/gradeable/{gradeable_id}/pdf")
      */
-    public function showStudentPDF($gradeable_id, $filename, $path) {
+    public function showStudentPDF($gradeable_id, $filename, $path, $grader = null) {
```
Did the changes to this file sneak in from another PR, or do they apply here?

jchulton replied:
This is how, when a user selects a certain peer in the peer feedback tab, only pdf annotations from that peer are rendered. I think it makes sense to apply them here.

emaicus commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -3178,6 +3178,25 @@ public function clearPeerGradingAssignments($gradeable_id) {
+     * @param string $feedback_full
+     * @param string $feedback_id
+     */
+    public function insertPeerGradingFeedback($grader, $student, $gradeable_id, $feedback_full, $feedback_id) {
+        $this->course_db->query("SELECT feedback_full FROM peer_feedback WHERE grader_id = ? AND user_id = ? AND g_id = ?", [$grader, $student, $gradeable_id]);
+        if (count($this->course_db->rows()) > 0) {
+            $this->course_db->query("UPDATE peer_feedback SET feedback_full = ?, feedback_id = ? WHERE grader_id = ? AND user_id = ? AND g_id = ?", [$feedback_full, $feedback_id, $grader, $student, $gradeable_id]);
+        }
+        else {
+            $this->course_db->query("INSERT INTO peer_feedback(grader_id, user_id, g_id, feedback_full, feedback_id) VALUES (?,?,?,?,?)", [$grader, $student, $gradeable_id, $feedback_full, $feedback_id]);
+        }
```
I think you can simplify this logic by adding an `ON CONFLICT` clause to your `sql` command. That way, you don't need to run the `SELECT` first, just do the insert. Use the `ON CONFLICT` to do an update.

emaicus commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -3204,6 +3223,21 @@ public function getPeerGradingAssignment($gradeable_id) {
+     *
+     * @param string $gradeable_id
+     */
+    public function getPeerFeedback($gradeable_id) {
```
Let's rename this `getAllPeerFeedback` or equivalent. I initially thought it would get the peer feedback for a specific grader based on the name.

emaicus commented on site/app/models/gradeable/Gradeable.php:
```diff
@@ -491,12 +491,42 @@ public function setPeerGradersList($input) {
+    
+    public function setPeerFeedback($grader_id, $student_id, $feedback_full, $feedback_id) {
+        $bad_input = [];
+        if ($this->core->getQueries()->getUserById($grader_id) == null) {
```
Use `===` for null checks.

jchulton replied:
Changed to ===

emaicus commented on site/app/models/gradeable/Gradeable.php:
```diff
@@ -491,12 +491,42 @@ public function setPeerGradersList($input) {
+        if ($this->core->getQueries()->getUserById($student_id) == null) {
+            array_push($bad_input, ($student_id));
+        }
+        if (!empty($bad_input)) {
+            $msg = "The given user id is not valid: ";
+            array_walk($bad_input, function ($val) use (&$msg) {
+                $msg .= " {$val}";
+            });
+            $this->core->addErrorMessage($msg);
+        }
```
Will this message reveal the user_id of a grader to a peer and break anonymization?

jchulton replied:
It will only show non-existent ids. I guess someone could guess a user_id? Should I change the message to not include the id?

emaicus replied:
Re-opening. Will this happen if the grader has left the course?

emaicus commented on site/app/models/gradeable/Gradeable.php:
```diff
@@ -491,12 +491,42 @@ public function setPeerGradersList($input) {
+        else {
+            $this->core->getQueries()->insertPeerGradingFeedback($grader_id, $student_id, $this->getId(), $feedback_full, $feedback_id);
+        }
+        var_dump($this->core->getQueries()->getPeerFeedback($this->getId()));
```
Remove before merging.

jchulton replied:
Yep, done. Sorry

emaicus commented on site/app/models/gradeable/Gradeable.php:
```diff
@@ -491,12 +491,42 @@ public function setPeerGradersList($input) {
+    }
+    
+    public function getPeerFeedback($grader_id, $anon_id) {
+        $feedback = $this->core->getQueries()->getPeerFeedback($this->getId());
```
Perhaps add a database function `getPeerFeedback($user_id, $is_team)` which returns the results for a single team/user_id.

jchulton replied:
Added this function: getPeerFeedbackInstance

emaicus commented on site/app/templates/autograding/PeerResults.twig:
```diff
@@ -9,6 +9,38 @@ $( document ).ready(function() {
+function open_peer_feedback_tab(user, ordered_graders) {
+    $('.page-content').hide();
+    $('#peer-feedback-'+user).show();
+    for (var i = 0 ; i<{{ordered_graders|json_encode|raw}}.length;i++){
```
Can you explain exactly what is going on here with your filters?

jchulton replied:
Added Comment explaining

emaicus commented on site/app/controllers/grading/ElectronicGraderController.php:
```diff
@@ -2373,4 +2373,27 @@ private function getStats(Gradeable $gradeable, User $grader, bool $full_stats,
         }
     }
+    
+    /**
+     * @Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/feedback/set", methods={"POST"})
+     */
+    public function ajaxSetPeerFeedback($gradeable_id) {
```
Can we add access control here to make certain we are allowed to update this feedback?

jchulton replied:
The blank return statemetns in the function should act as access control: ending the function early if the user doesn't have permission to update feedback

jchulton commented:
I've changed the select feedback buttons and added a saving/saved indicator, as below:
![image](https://user-images.githubusercontent.com/11095297/86084282-70ccef80-ba6a-11ea-8678-0d31a31ac1bd.png)

emaicus commented on migration/migrator/data/course_tables.sql:
```diff
@@ -417,6 +417,21 @@ CREATE TABLE peer_assign (
+    user_id character varying(255),
+    team_id character varying(255),
+    g_id character varying(255) NOT NULL,
+    feedback_full character varying(255),
+    feedback_id character varying(255),
```
It looks like you dropped one of these in the migration but forgot to do it here.

jchulton replied:
Dropped a column to match the migration file

emaicus commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -3249,15 +3249,32 @@ public function removePeerAssignmentsForGrader($gradeable_id, $grader_id) {
+     * @param string $gradeable_id
+     * @param string $feedback
      */
+    public function insertPeerGradingFeedback($grader, $student, $gradeable_id, $feedback) {
+        $this->course_db->query("SELECT feedback FROM peer_feedback WHERE grader_id = ? AND user_id = ? AND g_id = ?", [$grader, $student, $gradeable_id]);
+        if (count($this->course_db->rows()) > 0) {
+            $this->course_db->query("UPDATE peer_feedback SET feedback = ? WHERE grader_id = ? AND user_id = ? AND g_id = ?", [$feedback, $grader, $student, $gradeable_id]);
+        }
+        else {
+            $this->course_db->query("INSERT INTO peer_feedback(grader_id, user_id, g_id, feedback) VALUES (?,?,?,?)", [$grader, $student, $gradeable_id, $feedback]);
+        }
+    }
+  
+  /**
+   * Bulk Uploads Peer Grading Assignments
+   *
+   * @param string $values
+   */
```
This doesn't appear to work for teams.

jchulton replied:
As discussed below, this will use the user_id of the user giving feedback, not team_id

emaicus commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -3283,6 +3300,29 @@ public function getPeerGradingAssignment($gradeable_id) {
+     * @param string $gradeable_id
+     */
+    public function getPeerFeedback($gradeable_id) {
+        $this->course_db->query("SELECT grader_id, user_id, feedback FROM peer_feedback WHERE g_id = ? ORDER BY grader_id", [$gradeable_id]);
```
Same as above, you seem to be `user_id` here without considering `team_id`. But now that I think of it, what does it mean for a team to give a grader feedback? Is that actually how we want this to work?

jchulton replied:
As discussed, we will use the user_id of the student giving feedback

emaicus commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -3283,6 +3300,29 @@ public function getPeerGradingAssignment($gradeable_id) {
+        return $return;
+    }
+    
+    public function getPeerFeedbackInstance($gradeable_id, $grader_id, $user_id) {
+        $this->course_db->query("SELECT feedback FROM peer_feedback WHERE g_id = ? AND grader_id = ? AND user_id = ? ORDER BY grader_id", [$gradeable_id, $grader_id, $user_id]);
+        $results = $this->course_db->rows();
+        if (count($results) > 0) {
+            return $results[0]['feedback'];
+        }
+        return null;
+    }
```
Same

jchulton replied:
As above

emaicus commented on site/app/templates/autograding/PeerResults.twig:
```diff
@@ -18,6 +20,27 @@ function open_peer_feedback_tab(user, ordered_graders) {
     $('#peer_feedback_nav_'+user).addClass('active-btn');
 }
+function set_peer_feedback(grader_id, user_id, g_id, feedback) {
+    $('#peer_feedback_no_response_'+grader_id).css("background-color", "#ffffff");
+    $('#peer_feedback_thanks_'+grader_id).css("background-color", "#ffffff");
+    $('#peer_feedback_helpful_'+grader_id).css("background-color", "#ffffff");
+    $('#peer_feedback_detailed_'+grader_id).css("background-color", "#ffffff");
+    $('#peer_feedback_inappropriate_'+grader_id).css("background-color", "#ffffff");
+    $('#peer_feedback_no_response_'+grader_id).css("color", "#008000");
+    $('#peer_feedback_thanks_'+grader_id).css("color", "#008000");
+    $('#peer_feedback_helpful_'+grader_id).css("color", "#008000");
+    $('#peer_feedback_detailed_'+grader_id).css("color", "#008000");
+    $('#peer_feedback_inappropriate_'+grader_id).css("color", "#800000");
+    $('#peer_feedback_'+feedback+'_'+grader_id).css("color", "#ffffff");
```
Could this just be include in the css for this page?

jchulton replied:
Added to server.css

emaicus commented on site/app/models/gradeable/Gradeable.php:
```diff
@@ -530,12 +530,50 @@ public function setPeerGradersList($input) {
+        }
+    }
+    
+    public function getPeerFeedback($grader_id, $anon_id) {
+        $user_id = $this->core->getQueries()->getUserFromAnon($anon_id)[$anon_id];
+        $feedback = $this->core->getQueries()->getPeerFeedbackInstance($this->getId(), $grader_id, $user_id);
+        if ($feedback == 'thanks') {
+            return 'Thank you!';
+        }
+        elseif ($feedback == 'helpful') {
+            return 'This feedback was helpful to me!';
+        }
+        elseif ($feedback == 'detailed') {
+            return 'This feedback was detailed, specific, and/or technical';
+        }
+        elseif ($feedback == 'inappropriate') {
+            return 'This feedback was inaccurate and/or inappropriate';
+        }
+        return 'No response';
+    }
```
I don't think this function is ever called.

jchulton replied:
It is called in RubricPanel.twig in the following line 

Peer Feedback: {{gradeable.getPeerFeedback(grader_id, anon_id)}}. This is where we translate the database entry to twig output, since we are no longer storing both in the table

emaicus commented on site/app/templates/autograding/PeerResults.twig:
```diff
@@ -330,9 +344,25 @@ function switch_feedback_modes() {
+            <b>Peer Feedback:</b>
+            <div>
+                </br>
+                <a class="btn cat-buttons" data-color="#008000" style="border-color: #008000; color: #008000;" id="peer_feedback_no_response_{{grader}}" onclick="set_peer_feedback('{{grader}}', '{{user_id}}', '{{gradeable_id}}','no_response', '>No Response');">No Response</a>
+                <a class="btn cat-buttons" data-color="#008000" style="border-color: #008000; color: #008000;" id="peer_feedback_thanks_{{grader}}" onclick="set_peer_feedback('{{grader}}', '{{user_id}}', '{{gradeable_id}}', 'thanks', 'Thank You!');">Thank You</a>
+                <a class="btn cat-buttons" data-color="#008000" style="border-color: #008000; color: #008000;" id="peer_feedback_helpful_{{grader}}" onclick="set_peer_feedback('{{grader}}', '{{user_id}}', '{{gradeable_id}}','helpful', 'This feedback was helpful to me');">This feedback was helpful to me!</a>
+                <a class="btn cat-buttons" data-color="#008000" style="border-color: #008000; color: #008000;" id="peer_feedback_detailed_{{grader}}" onclick="set_peer_feedback('{{grader}}', '{{user_id}}', '{{gradeable_id}}', 'detailed', 'This feedback was detailed, specific, and/or technical');">This feedback was detailed, specific, and/or technical</a>
+                <a class="btn cat-buttons" data-color="#800000" style="border-color: #800000; color: #800000;" id="peer_feedback_inappropriate_{{grader}}" onclick="set_peer_feedback('{{grader}}', '{{user_id}}', '{{gradeable_id}}', 'inappropriate', 'This feedback was inaccurate and/or inappropriate');">This feedback was inaccurate and/or inappropriate</a>
```
Same as above, can we get the styles into a css stylesheet?

jchulton replied:
Added into server.cs

emaicus commented on site/public/js/server.js:
```diff
@@ -1624,6 +1624,36 @@ function enableKeyToClick(){
+        contentType: false,
+        success: function(data) {
+            try {
+                console.log(data);
```
stray print statement?

jchulton replied:
Removed the print

emaicus commented on site/app/templates/autograding/PeerResults.twig:
```diff
@@ -88,20 +90,21 @@ function switch_feedback_modes() {
-                {% endif %}                
-            {% endfor %}
-            </div>
+            <div id="normal_peer_wrapper">
+                {% for component in peer_components %}
+                    {% if component.peer_ids | length <= 5 %}
+                        {{ self.renderFewPeerResults(component, peer_aliases, ordered_graders)  }}
+                    {% else %}
+                        {{ self.renderManyPeerResults(component) }}
+                    {% endif %}                
+                {% endfor %}
+                </div>
```
```suggestion
            <div id="normal_peer_wrapper">
                {% for component in peer_components %}
                    {% if component.peer_ids | length <= 5 %}
                        {{ self.renderFewPeerResults(component, peer_aliases, ordered_graders)  }}
                    {% else %}
                        {{ self.renderManyPeerResults(component) }}
                    {% endif %}                
                {% endfor %}
            </div>
```

jchulton replied:
Committed the above change

emaicus commented on migration/migrator/data/course_tables.sql:
```diff
@@ -417,6 +417,20 @@ CREATE TABLE peer_assign (
+--
+-- Name: peer_feedback; Type: TABLE; Schema: public; Owner: -
+--
+
+CREATE TABLE peer_feedback (
+    pf_id integer NOT NULL,
+    grader_id character varying(255) NOT NULL,
+    user_id character varying(255),
+    team_id character varying(255),
+    g_id character varying(255) NOT NULL,
+    feedback character varying(255),
+    CONSTRAINT user_team_id_check CHECK (user_id IS NOT NULL OR team_id IS NOT NULL)
+);
```
See my previous review for how I think we should restructure this table. (Remove `pf_id` and `team_id`).

jchulton replied:
Removed pf_id and team_id

emaicus commented on migration/migrator/data/course_tables.sql:
```diff
@@ -1167,6 +1181,44 @@ ALTER TABLE ONLY peer_assign
+ALTER TABLE ONLY peer_feedback
+            ADD CONSTRAINT peer_feedback_team_id_fkey FOREIGN KEY (team_id) REFERENCES gradeable_teams(team_id) ON DELETE CASCADE;
+
+CREATE SEQUENCE IF NOT EXISTS peer_feedback_pf_id_seq
+            START WITH 1
+            INCREMENT BY 1
+            NO MINVALUE
+            NO MAXVALUE
+            CACHE 1;
+            
+
+ALTER SEQUENCE peer_feedback_pf_id_seq OWNED BY peer_feedback.pf_id;
+ALTER TABLE ONLY peer_feedback ALTER COLUMN pf_id SET DEFAULT nextval('peer_feedback_pf_id_seq'::regclass);
```
Can be removed if we get rid of `team_id` and `pf_seq`.

jchulton replied:
Removed team_id pf_seq

emaicus commented on migration/migrator/migrations/course/20200608034200_peer_feedback.py:
```diff
@@ -0,0 +1,110 @@
+    )
+
+    # Create overall comment table
+    database.execute(
+        """
+       CREATE TABLE IF NOT EXISTS peer_feedback (
+            pf_id integer NOT NULL,
+            grader_id character varying(255) NOT NULL,
+            user_id character varying(255),
+            team_id character varying(255),
+            g_id character varying(255) NOT NULL,
+            feedback character varying(255),
+            CONSTRAINT user_team_id_check CHECK (user_id IS NOT NULL OR team_id IS NOT NULL)
+        );
+        """
+    )
```
Make the changes mentioned above to the migration as well

jchulton replied:
Made the above changed

emaicus commented on migration/migrator/migrations/course/20200608034200_peer_feedback.py:
```diff
@@ -0,0 +1,110 @@
+
+
+    database.execute("ALTER TABLE peer_feedback DROP CONSTRAINT IF EXISTS peer_feedback_pkey")
+    database.execute(
+        """
+        ALTER TABLE ONLY peer_feedback
+            ADD CONSTRAINT peer_feedback_pkey PRIMARY KEY (pf_id);
+        """
+    )
```
Make the changes mentioned above to the migration as well

jchulton replied:
Made the mentioned changes (removing pf_id and team_id)

emaicus commented on migration/migrator/migrations/course/20200608034200_peer_feedback.py:
```diff
@@ -0,0 +1,110 @@
+        """
+    )
+
+    database.execute(
+        """
+        CREATE SEQUENCE IF NOT EXISTS peer_feedback_pf_id_seq
+            START WITH 1
+            INCREMENT BY 1
+            NO MINVALUE
+            NO MAXVALUE
+            CACHE 1;
+        """)
+
+    database.execute("ALTER SEQUENCE peer_feedback_pf_id_seq OWNED BY peer_feedback.pf_id;")
+    database.execute("ALTER TABLE ONLY peer_feedback ALTER COLUMN pf_id SET DEFAULT nextval('peer_feedback_pf_id_seq'::regclass);")
```
Make the changes mentioned above to the migration as well

jchulton replied:
Made the mentioned changes (removing pf_id and team_id)

emaicus commented on site/app/views/grading/ElectronicGraderView.php:
```diff
@@ -1218,6 +1224,8 @@ public function renderRubricPanel(GradedGradeable $graded_gradeable, int $displa
 
         return $return . $this->core->getOutput()->renderTwigTemplate("grading/electronic/RubricPanel.twig", [
                 "showNewInterface" => $showNewInterface,
+                "gradeable" => $gradeable,
+                "anon_ids" => $anon_ids,
                 "gradeable_id" => $gradeable->getId(),
                 "is_ta_grading" => $gradeable->isTaGrading(),
                 "anon_id" => $graded_gradeable->getSubmitter()->getAnonId(),
```
I see that there is both an `anon_id` and an `anon_ids` here. Do we need both?

jchulton replied:
Yes, anon_id represents the anon_id of the student/team. Anon_ids represents the anon_ids of all the students in the team. Renamed to student_anon_ids for clarity

emaicus commented on site/public/js/server.js:
```diff
@@ -1624,6 +1624,36 @@ function enableKeyToClick(){
+        cache: false,
+        contentType: false,
+        success: function(data) {
+            console.log(data);
```
stray print.

jchulton replied:
Removed the stray print

jchulton commented on migration/migrator/data/course_tables.sql:
```diff
@@ -1933,6 +1944,26 @@ ALTER TABLE ONLY public.queue
+-- Name: peer_feedback_g_id_fkey; Type: FK CONSTRAINT; Schema: public; Owner: -
+--
+
+ALTER TABLE ONLY peer_feedback
+            ADD CONSTRAINT peer_feedback_g_id_fkey FOREIGN KEY (g_id) REFERENCES gradeable(g_id) ON DELETE CASCADE;
+
+--
+-- Name: peer_feedback_grader_id_fkey; Type: FK CONSTRAINT; Schema: public; Owner: -
+--
+
+ALTER TABLE ONLY peer_feedback
+            ADD CONSTRAINT peer_feedback_grader_id_fkey FOREIGN KEY (grader_id) REFERENCES users(user_id) ON DELETE CASCADE;
```
Added foreigh key below

jchulton commented on migration/migrator/migrations/course/20200608034200_peer_feedback.py:
```diff
@@ -0,0 +1,75 @@
+        """
+    )
+    
+    database.execute("ALTER TABLE peer_feedback DROP CONSTRAINT IF EXISTS peer_feedback_g_id_fkey")
+    database.execute(
+        """
+        ALTER TABLE ONLY peer_feedback
+            ADD CONSTRAINT peer_feedback_g_id_fkey FOREIGN KEY (g_id) REFERENCES gradeable(g_id) ON DELETE CASCADE;
+        """
+    )
+
+    database.execute("ALTER TABLE peer_feedback DROP CONSTRAINT IF EXISTS peer_feedback_user_id_fkey")
+    database.execute(
+        """
+        ALTER TABLE ONLY peer_feedback
+            ADD CONSTRAINT peer_feedback_user_id_fkey FOREIGN KEY (user_id) REFERENCES users(user_id) ON DELETE CASCADE;
+
+        """
+    )
+    
+    database.execute("ALTER TABLE peer_feedback DROP CONSTRAINT IF EXISTS peer_feedback_grader_id_fkey")
+    database.execute(
+        """
+        ALTER TABLE ONLY peer_feedback
+            ADD CONSTRAINT peer_feedback_grader_id_fkey FOREIGN KEY (grader_id) REFERENCES users(user_id) ON DELETE CASCADE;
```
Added creation of primary key

jchulton commented on site/app/models/gradeable/Gradeable.php:
```diff
@@ -530,12 +530,50 @@ public function setPeerGradersList($input) {
             $this->core->getQueries()->clearPeerGradingAssignment($this->getId());
             foreach ($input as $row_num => $vals) {
-                $this->core->getQueries()->insertPeerGradingAssignment($vals["grader"], $vals["student"], $this->getId());
+                $this->core->getQueries()->insertPeerGradingAssignment($vals["grader"], $vals["student"], $vals["feedback"]);
```
That should just be the grader_id. Changed back

jchulton commented on migration/migrator/data/course_tables.sql:
```diff
@@ -1933,6 +1944,26 @@ ALTER TABLE ONLY public.queue
+--
+
+ALTER TABLE ONLY peer_feedback
+            ADD CONSTRAINT peer_feedback_pkey PRIMARY KEY (pf_id);
```
Removed pf_id

bmcutler commented:
@jchulton please resolve conflicts
@emaicus please review code afterwards

emaicus commented on site/app/controllers/grading/ElectronicGraderController.php:
```diff
@@ -2493,6 +2493,38 @@ public function ajaxSetPeerFeedback($gradeable_id) {
         }
         $gradeable->setPeerFeedback($grader_id, $user_id, $feedback);
         $this->core->getOutput()->renderJsonSuccess("Feedback successfully uploaded");
```
It looks like this used to return true before you fixed your merge conflicts.

jchulton replied:
Readded the return true

jchulton commented:
Tested and functional
